src/events/event_manager.h

```diff
@@ -0,0 +1,183 @@
+        * @fn                    RemoveAllCallbacksByType
+        * @brief                 Templated function which removes all event handlers for a specified event
+        */
+
+        template<class T>
+        static void RemoveAllCallbacksByType() noexcept;
+
+        /**
+        * @fn                    Flush
+        * @brief                 This function runs callbacks for all events that are currently in an event queue
+        * @param [int]           n -> Number of events you want to proceed. If n == -1 (or argumeter left blank)
+        *                             this will iterate through whole queue
+        */
+
+        static void Flush(int n = -1) noexcept;
+
+        /**
+        * @fn                    Flush
+        * @brief                 This function should be called at the end of program to clean up allocated memory
+        */
+
+        static void CleanUp() noexcept;
+
+    private:
+
+        /**
+        * @fn                    Call
+        * @brief                 This function pops first element of the queue and runs all functions associated with that event
+        */
+
+        static inline void Call() noexcept;
+
+    public:
+        static std::unordered_map<std::type_index, std::vector<Handler*>*> m_Callbacks; /**< Unordered map of events' type indexes and associated functions */
+        static std::deque<Event*> m_EventQueue; /**< Basically an event queue */
+    }; // END OF CLASS EventManager
+
+    template<class T, class F, class... Args>
+    unsigned EventManager::Subscribe(HandleType type, F&& callback, Args&& ... args) {
+
+        static_assert(std::is_base_of<Event, T>::value, "Event is not the base of T");
+
+        if(m_Callbacks.find(typeid(T)) == m_Callbacks.end()) {
+            m_Callbacks[typeid(T)] = new std::vector<Handler*>;
+        }
+
+        switch(type) {
+            case HandleType::SYNC:
+                m_Callbacks[typeid(T)]->push_back(new FunctionHandler(callback, args...));
+                break;
+            case HandleType::ASYNC:
+                m_Callbacks[typeid(T)]->push_back(new AsyncFunctionHandler(callback, args...));
+                break;
+        }
+
+        return m_Callbacks[typeid(T)]->back()->GetID();
+    }
+
+    template<class T>
+    void EventManager::Unsubscibe(unsigned id) noexcept {
+        static_assert(std::is_base_of<Event, T>::value, "Event is not the base of T");
+        auto iter = m_Callbacks.find(typeid(T));
+
+        if(iter == m_Callbacks.end()) return;
+
+        auto vector = iter->second;
+
+        for(auto i = vector->begin(); i != vector->end(); ++i) {
+            if((**i) == id) {
+                delete *i;
+                vector->erase(i);
+                return;
+            }
+        }
+    }
+
+    template<class T, class... Args>
+    void EventManager::Emit(Args&& ... args) {
+        static_assert(std::is_base_of<Event, T>::value, "Event is not the base of T");
+
+        T* event = new T{args...};
```

Comment:
we should stick to one type of initialization.

Reply:
Then let's proceed with the newest standard

Reply:
:(